src/net/ossl.cc

```diff
@@ -740,61 +752,20 @@ class session : public enable_shared_from_this<session>, public session_impl {
           });
     }
 
-    // This function will attempt to read data out of the OpenSSL out_bio()
-    // which the SSL session writes to.  If any data is present, it will
-    // push it into the _out stream and save off the future into `_output_pending`.
-    // If there is data waiting to be sent, this function will wait for
-    // `_output_pending` to resolve.
-    future<> perform_push() {
-        return _output_pending.then([this] {
-            return repeat_until_value(
-                [this, msg = scattered_message<char>()] () mutable {
-                    using ret_t = std::optional<scattered_message<char>>;
-                    buf_type buf(BIO_ctrl_pending(out_bio()));
-                    auto n = BIO_read(
-                        out_bio(), buf.get_write(), buf.size());
-                    if (n > 0) {
-                        buf.trim(n);
-                        msg.append(std::move(buf));
-                    } else if (!BIO_should_retry(out_bio())) {
-                        _error = std::make_exception_ptr(
-                            ossl_error::make_ossl_error(
-                                "Failed to read from out_bio()"));
-                        return make_exception_future<ret_t>(_error);
-                    }
-                    if (BIO_ctrl_pending(out_bio()) == 0) {
-                        return make_ready_future<ret_t>(std::move(msg));
-                    }
-                    return make_ready_future<ret_t>();
-                }
-            ).then([this](scattered_message<char> msg) mutable {
-                if (msg.size() > 0) {
-                    _output_pending = _out.put(std::move(msg).release());
-                } else {
-                    _output_pending = make_ready_future();
-                }
-            });
+    future<>
+    handle_output_error(std::exception_ptr ptr) {
```

Comment:
nit: Maybe the PR could have been just the edits to handle_output_error , and associated modified code, then the additions to the new BIO_methods in later commits

Reply:
Part of the change to use the custom BIO changed how some of the data movement happened.